core/src/main/scala/org/apache/spark/crypto/CryptoOutputStream.scala

```diff
@@ -0,0 +1,224 @@
+  outBuffer = ByteBuffer.allocateDirect(bufferSize)
+  streamOffset = streamOffsetVal
+  try {
+    encryptor = codec.createEncryptor
+  }
+  catch {
+    case e: GeneralSecurityException => {
+      throw e
+    }
+  }
+  updateEncryptor
+
+
+  def this(out: OutputStream, codec: CryptoCodec, bufferSize: Int, key: Array[Byte],
+           iv: Array[Byte]) {
+    this(out, codec, bufferSize, key, iv, 0)
+  }
+
+  def this(out: OutputStream, codec: CryptoCodec, key: Array[Byte], iv: Array[Byte], streamOffset:
+  Long) {
+    this(out, codec, CryptoStreamUtils.getBufferSize, key, iv, streamOffset)
+  }
+
+  def this(out: OutputStream, codec: CryptoCodec, key: Array[Byte], iv: Array[Byte]) {
+    this(out, codec, key, iv, 0)
+  }
+
+  def getWrappedStream: OutputStream = {
+    out
+  }
+
+  /**
+   * Encryption is buffer based.
+   * If there is enough room in {@link #inBuffer}, then write to this buffer.
+   * If {@link #inBuffer} is full, then do encryption and write data to the
+   * underlying stream.
+   * @param b the data.
+   * @param offVal the start offset in the data.
+   * @param lenVal the number of bytes to write.
+   * @throws IOException
+   */
+  override def write(b: Array[Byte], offVal: Int, lenVal: Int) {
+    var off = offVal
+    var len = lenVal
+    checkStream
+    if (b == null) {
+      throw new NullPointerException
+    }
+    else if (off < 0 || len < 0 || off > b.length || len > b.length - off) {
+      throw new IndexOutOfBoundsException
+    }
+    while (len > 0) {
+      val remaining: Int = inBuffer.remaining
+      if (len < remaining) {
+        inBuffer.put(b, off, len)
+        len = 0
+      }
+      else {
+        inBuffer.put(b, off, remaining)
+        off += remaining
+        len -= remaining
+        encrypt
+      }
+    }
+  }
+
+  /**
+   * Do the encryption, input is {@link #inBuffer} and output is
+   * {@link #outBuffer}.
+   */
+  def encrypt {
+    Preconditions.checkState(inBuffer.position >= padding)
+    if (inBuffer.position != padding) {
+      inBuffer.flip
+      outBuffer.clear
+      encryptor.encrypt(inBuffer, outBuffer)
+      inBuffer.clear
+      outBuffer.flip
+      if (padding > 0) {
+        outBuffer.position(padding)
+        padding = 0
+      }
+      val len: Int = outBuffer.remaining
+      val tmp: Array[Byte] = getTmpBuf
+      outBuffer.get(tmp, 0, len)
+      out.write(tmp, 0, len)
+      streamOffset += len
+      if (encryptor.isContextReset) {
+        updateEncryptor
+      }
+    }
+  }
+
+  /** Update the {@link #encryptor}: calculate counter and {@link #padding}. */
+  def updateEncryptor {
+    val counter: Long = streamOffset / codec.getCipherSuite.algoBlockSize
+    padding = (streamOffset % codec.getCipherSuite.algoBlockSize).asInstanceOf[Byte]
+    inBuffer.position(padding)
+    codec.calculateIV(initIV, counter, iv)
+    encryptor.init(key, iv)
+  }
+
+  def getTmpBuf: Array[Byte] = {
+    if (tmpBuf == null) {
```

Comment:
Why not initialize this in the constructor? Lazy initialization doesn't seem to be buying anything (unless you never write anything to the stream).